### PR TITLE
Fail repair after additive change in cluster topology

### LINKED ISSUE #1367: Repairs Not progressing, when Incremental repairs weren't used and cluster has range movement
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=36766654)

**Reaper Version in use v 3.2.0**

Currently, for non-incremental repairs with one segment per node, Reaper is not resilient against changes in token ranges; any repair schedules created before a range change would subsequently fail consistently, resulting in continuous retries. In addition, these failing repairs seem to take consistent priority over repairs for other keyspaces, blocking progress completely.

image
image
image
image

**ERROR**

```2023-06-07T14:10:55.595Z: ERROR  [2023-06-07 14:10:54,386] [****:1cfdb910-df93-11ed-9996-a1697ff8c53a] i.c.j.ClusterFacade - [tokenRangeToEndpoint] no replicas found for token range io.cassandrareaper.core.Segment@41265603  2023-06-07T14:10:55.595Z: ERROR  [2023-06-07 14:10:54,387] [****:1cfdb910-df93-11ed-9996-a1697ff8c53a] i.c.s.RepairRunner - RepairRun FAILURE, scheduling retry  2023-06-07T14:10:55.595Z: java.lang.IllegalArgumentException: no hosts provided to connectAny 2023-06-07T14:10:55.595Z:     at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135) 2023-06-07T14:10:55.596Z:     at io.cassandrareaper.jmx.JmxConnectionFactory.connectAny(JmxConnectionFactory.java:135) 2023-06-07T14:10:55.596Z:     at io.cassandrareaper.jmx.ClusterFacade.connectImpl(ClusterFacade.java:885) 2023-06-07T14:10:55.596Z:     at io.cassandrareaper.jmx.ClusterFacade.connect(ClusterFacade.java:869) 2023-06-07T14:10:55.596Z:     at io.cassandrareaper.service.RepairRunner.startNextSegment(RepairRunner.java:472) 2023-06-07T14:10:55.596Z:     at io.cassandrareaper.service.RepairRunner.run(RepairRunner.java:235) 2023-06-07T14:10:55.596Z:     at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) 2023-06-07T14:10:55.596Z:     at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125) 2023-06-07T14:10:55.596Z:     at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57) 2023-06-07T14:10:55.596Z:     at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78) 2023-06-07T14:10:55.596Z:     at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241) 2023-06-07T14:10:55.596Z:     at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) 2023-06-07T14:10:55.596Z:     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) 2023-06-07T14:10:55.596Z:     at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) 2023-06-07T14:10:55.596Z:     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) 2023-06-07T14:10:55.596Z:     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) 2023-06-07T14:10:55.596Z:     at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) 2023-06-07T14:10:55.596Z:     at java.base/java.lang.Thread.run(Thread.java:829) 2023-06-07T14:10:00.009Z: ERROR  [2023-06-07 14:09:59,816] [****:a8424cd0-df92-11ed-9996-a1697ff8c53a] i.c.j.ClusterFacade - [tokenRangeToEndpoint] no replicas found for token range io.cassandrareaper.core.Segment@60b78a64  2023-06-07T14:10:00.009Z: ERROR  [2023-06-07 14:09:59,816] [****:a8424cd0-df92-11ed-9996-a1697ff8c53a] i.c.s.RepairRunner - RepairRun FAILURE, scheduling retry  2023-06-07T14:10:00.009Z: java.lang.IllegalArgumentException: no hosts provided to connectAny 2023-06-07T14:10:00.009Z:     at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135) 2023-06-07T14:10:00.009Z:     at io.cassandrareaper.jmx.JmxConnectionFactory.connectAny(JmxConnectionFactory.java:135) 2023-06-07T14:10:00.009Z:     at io.cassandrareaper.jmx.ClusterFacade.connectImpl(ClusterFacade.java:885) 2023-06-07T14:10:00.009Z:     at io.cassandrareaper.jmx.ClusterFacade.connect(ClusterFacade.java:869) 2023-06-07T14:10:00.009Z:     at io.cassandrareaper.service.RepairRunner.startNextSegment(RepairRunner.java:472) 2023-06-07T14:10:00.009Z:     at io.cassandrareaper.service.RepairRunner.run(RepairRunner.java:235) 2023-06-07T14:10:00.009Z:     at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) 2023-06-07T14:10:00.009Z:     at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125) 2023-06-07T14:10:00.009Z:     at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57) 2023-06-07T14:10:00.009Z:     at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78) 2023-06-07T14:10:00.009Z:     at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241) 2023-06-07T14:10:00.009Z:     at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) 2023-06-07T14:10:00.009Z:     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) 2023-06-07T14:10:00.009Z:     at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) 2023-06-07T14:10:00.009Z:     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) 2023-06-07T14:10:00.009Z:     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) 2023-06-07T14:10:00.062Z:     at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) 2023-06-07T14:10:00.062Z:     at java.base/java.lang.Thread.run(Thread.java:829)```

Code flow for the error,

**RepairRunner**

```
private void startNextSegment() throws ReaperException, InterruptedException {
  boolean scheduleRetry = true;

  // We want to know whether a repair was started,
  // so that a rescheduling of this runner will happen.
  boolean repairStarted = false;

  // We have an empty slot, so let's start new segment runner if possible.
  // When in sidecar mode, filter on ranges that the local node is a replica for only.
  LOG.info("Attempting to run new segment...");
  List<RepairSegment> nextRepairSegments
      = context.config.isInSidecarMode()
          ? ((IDistributedStorage) context.storage)
              .getNextFreeSegmentsForRanges(
                  repairRunId, localEndpointRanges)
          : context.storage.getNextFreeSegments(
              repairRunId);

  Optional<RepairSegment> nextRepairSegment = Optional.empty();
  Collection<String> potentialReplicas = new HashSet<>();
  for (RepairSegment segment : nextRepairSegments) {
    Map<String, String> potentialReplicaMap = this.repairRunService.getDCsByNodeForRepairSegment(
        cluster, segment.getTokenRange(), repairUnit.getKeyspaceName(), repairUnit);
    potentialReplicas = repairUnit.getIncrementalRepair()
        ? Collections.singletonList(segment.getCoordinatorHost())
        : potentialReplicaMap.keySet();
    JmxProxy coordinator = clusterFacade.connect(cluster, potentialReplicas);
    if (nodesReadyForNewRepair(coordinator, segment, potentialReplicaMap, repairRunId)) {
      nextRepairSegment = Optional.of(segment);
      break;
    }
  }
```

Since incremental repair is not in use, Reaper computes potentialReplicas as the keys returned from getDCsByNodeForRepairSegment(), which is based on the current ring state and on data stored when the schedule was created **(segment.getTokenRange())**

**RepairRunService**

```
Map<String, String> getDCsByNodeForRepairSegment(
    Cluster cluster,
    Segment segment,
    String keyspace,
    RepairUnit repairUnit) throws ReaperException {

  final int maxAttempts = 2;
  for (int attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      JmxProxy jmxConnection = clusterFacade.connect(cluster);
      // when hosts are coming up or going down, this method can throw an UndeclaredThrowableException
      Collection<String> nodes = clusterFacade.tokenRangeToEndpoint(cluster, keyspace, segment);
      Map<String, String> dcByNode = Maps.newHashMap();
      nodes.forEach(node -> dcByNode.put(node, EndpointSnitchInfoProxy.create(jmxConnection).getDataCenter(node)));
      if (repairUnit.getDatacenters().isEmpty()) {
        return dcByNode;
      } else {
        return dcByNode.entrySet().stream()
          .filter(entry -> repairUnit.getDatacenters().contains(entry.getValue()))
          .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
      }
    }
```

The mapping is computed partly based on the output of tokenRangeToEndpoint(), which attempts to find a node that owns the range completely enclosing the segment under repair

**ClusterFacade**

```
public List<String> tokenRangeToEndpoint(Cluster cluster, String keyspace, Segment segment) {
  Set<Map.Entry<List<String>, List<String>>> entries;
  try {
    entries = getRangeToEndpointMap(cluster, keyspace).entrySet();
  } catch (ReaperException e) {
    LOG.error("[tokenRangeToEndpoint] no replicas found for token range {}", segment, e);
    return Lists.newArrayList();
  }

  for (Map.Entry<List<String>, List<String>> entry : entries) {
    BigInteger rangeStart = new BigInteger(entry.getKey().get(0));
    BigInteger rangeEnd = new BigInteger(entry.getKey().get(1));
    if (new RingRange(rangeStart, rangeEnd).encloses(segment.getTokenRanges().get(0))) {
      return entry.getValue();
    }
  }
  LOG.error("[tokenRangeToEndpoint] no replicas found for token range {}", segment);
  LOG.debug("[tokenRangeToEndpoint] checked token ranges were {}", entries);
  return Lists.newArrayList();
}
```

With one segment per node, each Segment corresponds to a single token range. If any additive/lateral range movements occur, then for at least some stored segments, no single endpoint will completely enclose its range. So tokenRangeToEndpoint() will return an empty list, which results in getDCsByNodeForRepairSegment() to return an empty map, which ultimately results in an empty list of potential coordinators being passed to connectAny()

**RepairRunner - JmxConnectionFactory**

```
@VisibleForTesting
public final JmxProxy connectAny(Collection<Node> nodes) throws ReaperException {

  Preconditions.checkArgument(
      null != nodes && !nodes.isEmpty(), "no hosts provided to connectAny");

  List<Node> nodeList = new ArrayList<>(nodes);
  Collections.shuffle(nodeList);
```

resulting in the errors logged above

The cluster in use has no incremental repair. What is the right way to proceed here, does it make sense to special case use cases involving range movement, in the same way that incremental repair is special cased.

┆Issue is synchronized with this [Jira Story](https://datastax.jira.com/browse/REAP-33) by [Unito](https://www.unito.io)
┆Issue Number: REAP-33

### PR DESCRIPTION
**Pull request by andresbeckruiz**
Token range movement occurs when a Cassandra cluster is expanded. When a cluster is expanded during an ongoing repair, repair segments that were previously calculated using old token ranges will no longer fully enclose each new token range. As a result, repairs hang as Reaper continuously retries failing segments that are no longer owned by a single Cassandra node.

This PR prevents repairs from hanging in this case by failing a repair early if there is an additive change in cluster topology. This PR also adds functionality to optionally reschedule a repair to run after a configurable amount of time. 

Fixes #1367

### CODE REVIEW

**Comment by andresbeckruiz** [comment]
I compiled and sucessfully ran all server tests, including five new tests I created in `RepairRunnerTopologyChangeTest.java` to verify my changes. During testing, I found that `manageIncRepairAbovePercentThresholdSchedule` in `SchedulingManagerTest.java` was flaky with and without my code changes, but was able to get the test to pass for both cases. 

Screenshot 2024-11-06 at 5 03 54 PM

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java [hunk 1]
```diff
@@ -245,12 +247,9 @@ public void testHangingRepair() throws InterruptedException, ReaperException, JM
...
     RepairUnit cf = storage.getRepairUnitDao().addRepairUnit(
-        RepairUnit.builder()
-            .clusterName(cluster.getName())
-            .keyspaceName("reaper")
-            .columnFamilies(cfNames)
-            .incrementalRepair(false)
-            .subrangeIncrementalRepair(false)
+        RepairUnit.builder().clusterName(cluster.getName())
+            .keyspaceName("reaper").columnFamilies(cfNames)
+            .incrementalRepair(false).subrangeIncrementalRepair(false)
```
Comment (adejanovski) [inline src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java]: **Todo:** Could you format these lines back the way they were?
It's not matching the style we use in the project.

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java [hunk 2]
```diff
@@ -402,10 +404,8 @@ public void testHangingRepairNewApi() throws InterruptedException, ReaperExcepti
     DateTimeUtils.setCurrentMillisFixed(timeRun);
     RepairUnit cf = storage.getRepairUnitDao().addRepairUnit(
         RepairUnit.builder()
-            .clusterName(cluster.getName())
-            .keyspaceName(ksName)
-            .columnFamilies(cfNames)
-            .incrementalRepair(incrementalRepair)
+            .clusterName(cluster.getName()).keyspaceName(ksName)
+            .columnFamilies(cfNames).incrementalRepair(incrementalRepair)
```
Comment (adejanovski) [inline src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java]: **Todo:** Could you format these lines back the way they were as well?

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java [hunk 3]
```diff
@@ -420,7 +420,7 @@ public void testHangingRepairNewApi() throws InterruptedException, ReaperExcepti
         Collections.singleton(
             RepairSegment.builder(
                 Segment.builder()
-                    .withTokenRange(new RingRange(BigInteger.ZERO, new BigInteger("100")))
+                    .withTokenRange(new RingRange(BigInteger.ZERO, new BigInteger("50")))
```
Comment (adejanovski) [inline src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java:429]: **Question:** what is the purpose of this change?

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java (hunk 1 above)
Comment (adejanovski) [reply src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java]: Was this a way to work around the checkstyle rule that limits the number of lines in a method?

**Review by adejanovski — COMMENTED** [review]
Successfully tested, that's a great improvement 👍 
I just have a couple questions around some formatting changes.

Could you rebase this PR on top of master to get the latest changes? I want to make sure the tests still pass with what's on master now.
Thanks!

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java (hunk 3 above)
Comment (andresbeckruiz) [reply src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java:429]: Previously, `testHangingRepair` and `testHangingRepairNewApi` simulated a hanging repair by recreating the bug that this patch fixes: constructing a repair segment with a token range (0-100) that is not fully enclosed by the originally calculated segments (represented in the [`sixNodeCluster` range map](https://github.com/thelastpickle/cassandra-reaper/blob/master/src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java#L147)). After introducing my patch which fails a repair early if this edge case is reached, the repairs were no longer hanging, causing both of these tests to fail.

After modifying the repair segment range to be 0-50 so that it is fully enclosed by the previously calculated segment with range 0-50, these tests now correctly simulate a hanging repair and pass. I confirmed that my changes to these tests do not affect any existing behavior by successfully running the updated tests before I added my patch. I further verified that the test behavior of triggering a repair twice is not affected by including:
````
verify(jmx, times(2)).triggerRepair(
        any(), any(), any(), any(), any(), any(), any(), anyInt()
);
````

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java (hunk 1 above)
Comment (andresbeckruiz) [reply src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java]: Yes, I was trying to work around the `MethodLength` checkstyle rule that limits methods to 150 lines and was preventing me from compiling the project. 

I added a suppression for this checkstyle rule on test files so that I could revert to the old formatting. Please let me know if I should keep the current formatting instead.

**Comment by andresbeckruiz** [comment]
Addressed comments, rebased, and confirmed that tests pass with newest changes. This is ready for another review when you get the chance, thank you @adejanovski

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java (hunk 3 above)
Comment (adejanovski) [reply src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java:429]: sounds good, thanks for the explanation!

File: src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java (hunk 1 above)
Comment (adejanovski) [reply src/server/src/test/java/io/cassandrareaper/service/RepairRunnerHangingTest.java]: I'm happy with the suppression, that rule is quite annoying for tests

File: src/server/src/checkstyle/suppressions.xml [hunk 4]
```diff
@@ -20,5 +20,6 @@
   <!-- suppressions for tests -->
   <suppress checks="MethodName" files="[/\\]src[/\\]test[/\\]java[/\\]" />
   <suppress checks="AbbreviationAsWordInName" files="[/\\]src[/\\]test[/\\]java[/\\]" />
+  <suppress checks="MethodLength" files="[/\\]src[/\\]test[/\\]java[/\\]" />
```
Comment (adejanovski) [inline src/server/src/checkstyle/suppressions.xml:23]: Neat!

**Review by adejanovski — APPROVED** [review]

**Comment by vikingUnet** [comment]
Thanks for fix! Can you please add this fix to CHANGELOG.md ? It is still have 3.7 version as latest =( Where i can download deb package for trying new version?
